libraries/botbuilder-dialogs/src/prompts/oauthPrompt.ts

```diff
@@ -320,10 +322,33 @@ export class OAuthPrompt extends Dialog {
         await context.sendActivity(msg);
     }
 
-    private async recognizeToken(context: TurnContext): Promise<PromptRecognizerResult<TokenResponse>> {
+    private async recognizeToken(dc: DialogContext): Promise<PromptRecognizerResult<TokenResponse>> {
+        const context = dc.context;
         let token: TokenResponse|undefined;
         if (this.isTokenResponseEvent(context)) {
             token = context.activity.value as TokenResponse;
+
+            // Fix-up the DialogContext's state context if this was received from a skill host caller.
+            const state: CallerInfo = dc.activeDialog.state[this.PersistedCaller];
+            if (state !== null) {
+                // Set the ServiceUrl to the skill host's Url
+                dc.context.activity.serviceUrl = state.callerServiceUrl;
+
+                // Recreate a ConnectorClient and set it in TurnState so replies use the correct one
+                if (!(typeof (context.adapter as any).createConnectorClientWithIdentity === 'function')) {
```

Comment:
so we're now doing this in  couple of places - suggests we have something missing in the core abstractions!

Reply:
Agreed, the botbuilder/Microsoft.Bot.Builder libraries are tightly coupled to the Connector code which is not always desirable. The Connector libraries tie the SDKs to HTTP, but the transport protocol isn't guaranteed to be HTTP with browsers, electron, mobile apps (e.g. Xamarin), IoT devices, or notebooks.

One area that could help in this decoupling is the introduction of a `"TransportClient"` interface similar to HttpClient in .netcore.

Reply:
One simple cleanup would be to use a type guard... [https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards)


---
In reply to: [418193184](https://github.com/microsoft/botbuilder-js/pull/2144#discussion_r418193184) [](ancestors = 418193184)

Reply:
stevenic and I synced offline, the type guard would still only result in a casting of any because the required interface is `botframework-connector` dependent and lives in `botbuilder`.